[WebXR] Fullscreen AR immersive mode

djg opened the pull request:
#### 481f557cdcb267e860b6563809cd6acee3be8126
<pre>
[WebXR] Fullscreen AR immersive mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=265297">https://bugs.webkit.org/show_bug.cgi?id=265297</a>
<a href="https://rdar.apple.com/118759729">rdar://118759729</a>

Reviewed by Dean Jackson.

Implement gesture to exit AR immersive mode and display content fullscreen.

To correctly exit RenderLoop thread with out removing backing state, termination
handling has been re-written to ensure that the thread has shutdown before
switching state. This removes the need for the Terminating state, which has been
removed.

* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.h:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::startSession):
(WebKit::ARKitCoordinator::endSessionIfExists):
(WebKit::ARKitCoordinator::endSession):
(WebKit::ARKitCoordinator::scheduleAnimationFrame):
(WebKit::ARKitCoordinator::submitFrame):
(WebKit::ARKitCoordinator::renderLoop):
(WebKit::ARKitCoordinator::currentSessionHasEnded): Deleted.
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.h:
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
(-[_WKARPresentationSession initWithSession:descriptor:]):
(-[_WKARPresentationSession viewDidLoad]):
(-[_WKARPresentationSession _cancelAction:]):
(-[_WKARPresentationSession _enterFullscreen]):
(-[_WKARPresentationSession _exitFullscreen]):

Canonical link: <a href="https://commits.webkit.org/271188@main">https://commits.webkit.org/271188@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1524a54feb3a9a50b1c3b2478e56eb4bfe01a983

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6290 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28898 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29872 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25275 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3683 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25039 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5046 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23718 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4380 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30512 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25247 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25152 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30669 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2716 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6066 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6636 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4996 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.h:
```diff
@@ -48,6 +48,7 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, nonnull, retain, readonly) id<MTLSharedEvent> completionEvent;
 @property (nonatomic, nullable, retain, readonly) id<MTLTexture> colorTexture;
 @property (nonatomic, readonly) NSUInteger renderingFrameIndex;
+@property (nonatomic, readonly) bool sessionEndRequested;
```
this is read on some queue, but written on the main thread, should this be an atomic property instead?

jyavenard commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -117,6 +122,7 @@ - (nullable instancetype)initWithSession:(ARSession *)session descriptor:(WKARPr
         _session = session;
         _sessionDescriptor = adoptNS([descriptor copy]);
         _device = adoptNS(MTLCreateSystemDefaultDevice());
+        _sessionEndRequested = false;
```
if the property is to be made atomic, this would have to become self.sessionEndRequested

jyavenard commented on Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
```diff
@@ -233,17 +256,35 @@ - (void)viewWillDisappear:(BOOL)animated
 
+- (void)_cancelAction:(id)sender
+{
+    _sessionEndRequested = true;
```
same here (if atomic)

jyavenard commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.h:
```diff
@@ -72,12 +72,13 @@ class ARKitCoordinator final : public PlatformXRCoordinator {
+        // NOTE: Since we only require atomic load/store, does this need to be
+        // atomic? Making terminateRequested atomic breaks the ability to easily
+        // construct Active inside std::variant.
+        bool terminateRequested;
```
an std::atomic would be easier IMHO.

djg replied:
It doesn't compile with `std::atomic` since that can't be moved.

jyavenard commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
```diff
@@ -172,11 +190,12 @@
                 return;
             }
 
+            if (WTF::atomicLoad(&active.terminateRequested)) {
```
so then you can just do `if (active.terminateRequested)`

jyavenard commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
```diff
@@ -194,13 +213,15 @@
                 return;
             }
 
+            if (WTF::atomicLoad(&active.terminateRequested)) {
```
same here

jyavenard commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
```diff
@@ -148,14 +148,32 @@
+    RELEASE_ASSERT(std::holds_alternative<Active>(m_state));
+    Active& active = std::get<Active>(m_state);
+
+    WTF::atomicStore(&active.terminateRequested, true);
```
and active.terminateRequested = true; 

Though, would that cause issue in the Variant definition?

kkinnunen-apple commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
```diff
@@ -238,6 +244,16 @@
             break;
 
         auto& active = *maybeActive;
+        if (WTF::atomicLoad(&active.terminateRequested))
```
Unsolicited feedback :) 
Above, you already switch on a shared variable completely unsynchronized. It's a bit convoluted.
What's the point of ensuring reading `active.terminateRequested` correct, if the whole premise is relying on getting the read for the get_if correct? And if the Active state wasn't correct, how could anything else be. Your synchronization state, semaphores etc. are in the m_state, and the way code is written it is as if it can be destroyed at any given moment.

It would be simpler to understand if you didn't access the "Active/Idle" as a variant in the render loop.
In fact, the ARKiitCoordinatior likely shouldn't be accessed in the render loop.

Something like

```
void arKitCoordinatorRenderLoop(Ref<RenderParameters> params) {
    while(!params->shouldStop.load()) {
     ....
    }
}
```

And whether the RenderParameters  is refcounted or just plain & depends on the mode if upon stopping you want to wait until the renderloop has really ended or not.
(RenderParameters is a poor name)

djg replied:
> What's the point of ensuring reading active.terminateRequested correct,

I didn't want the compiler to optimise out the check. I'm effectively using the atomic load to force re-reading the variable. (I could have used `volatile bool`, I suppose).

kkinnunen-apple commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
```diff
@@ -125,14 +125,14 @@
                 .renderThread = Thread::create("ARKitCoordinator session renderer", [this] { renderLoop(); }),
                 .renderSemaphore = Box<BinarySemaphore>::create(),
+                .terminateRequested = false,
             };
```
So this could be like 

```
struct RenderParameters {
        BinarySemaphore produceFrameSemaphore;
        std::atomic<bool> shouldQuit;
        ... sessionEventClient;
        ... presentationSession;
        ....;   
  };
  Box<RenderParameters> params = new RenderParameters {
            ....= WTFMove(sessionEventClient),
            ....= WTFMove(presentationSession),
  };


   m_state = Active {
                .pageIdentifier = page.webPageID(),
                .params = params;  
                .renderThread = Thread::create("ARKitCoordinator session renderer", [this, params] { arKitCoordinatorRenderLoop(*params); }),
             };
```

kkinnunen-apple commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
```diff
@@ -238,6 +244,16 @@
+            break;
+
+        if ([active.presentationSession sessionEndRequested]) {
+            callOnMainRunLoop([this]() {
```
these sort of cross-thread `this` captures cannot be made as I don't think ARKitCoordinator is thread-safe refcounted. So it might already be destroyed at the time the callback gets run.

djg replied:
The ARKitCoordinator is a `static LazyNeverDestroyed<ARKitCoordinator> xrCoordinator;` so I hope that it's never destroyed at the time the callback gets run.

kkinnunen-apple commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
```diff
@@ -238,6 +244,16 @@
+
+        if ([active.presentationSession sessionEndRequested]) {
+            callOnMainRunLoop([this]() {
+                endSession();
```
also running the endsession itself is incorrect, as that can be run simulataneusly by the other side (endSessionIfExists). So then you try to end a session that has already ended and this will assert in your asserts.

djg replied:
OK, while I think this happening is unlikely, I've reworked the code to check for an existing session in all cases.

kkinnunen-apple commented on Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
```diff
@@ -119,43 +126,60 @@
 
-            if (active.onFrameUpdate)
-                active.onFrameUpdate({ });
+            if (active.renderState->terminateRequested.exchange(true))
```
is this the only place that writes to terminateRequested? Is this re-entrant from multiple concurrent callers? It looks a bit odd to guard the variable this way if there is only one correct caller and it only flips once..

grorg reviewed (approved):
I think I'm confused by GitHub. I'm still seeing comments from Kimmo and JY even though you force-pushed. I will mark as r+ assuming you'll address them.

djg committed:
[WebXR] Fullscreen AR immersive mode
https://bugs.webkit.org/show_bug.cgi?id=265297
rdar://118759729

Reviewed by Dean Jackson.

Implement gesture to exit AR immersive mode and display content fullscreen.

To correctly exit RenderLoop thread with out removing backing state, termination
handling has been re-written to ensure that the thread has shutdown before
switching state. This removes the need for the Terminating state, which has been
removed.

* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.h:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::startSession):
(WebKit::ARKitCoordinator::endSessionIfExists):
(WebKit::ARKitCoordinator::endSession):
(WebKit::ARKitCoordinator::scheduleAnimationFrame):
(WebKit::ARKitCoordinator::submitFrame):
(WebKit::ARKitCoordinator::renderLoop):
(WebKit::ARKitCoordinator::currentSessionHasEnded): Deleted.
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.h:
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
(-[_WKARPresentationSession initWithSession:descriptor:]):
(-[_WKARPresentationSession viewDidLoad]):
(-[_WKARPresentationSession _cancelAction:]):
(-[_WKARPresentationSession _enterFullscreen]):
(-[_WKARPresentationSession _exitFullscreen]):

Canonical link: https://commits.webkit.org/271188@main
